File: gotest/leetcode/Leet1425_ConstrainedSubsequenceSum.go
package main

import (
	"fmt"
	"container/list"
)

// https://leetcode.com/problems/constrained-subsequence-sum/

// Given an integer array nums and an integer k, return the maximum sum of a non-empty
// subsequence of that array such that for every two consecutive integers in
// the subsequence, nums[i] and nums[j], where i<j, the condition j-i<=k is satisfied.
// A subsequence of an array is obtained by deleting some number of elements (can be zero)
// from the array, leaving the remaining elements in their original order.
// Example 1:
//   Input: nums = [10,2,-10,5,20], k = 2
//   Output: 37
//   Explanation: The subsequence is [10, 2, 5, 20].
// Example 2:
//   Input: nums = [-1,-2,-3], k = 1
//   Output: -1
//   Explanation: The subsequence must be non-empty, so we choose the largest number.
// Example 3:
//   Input: nums = [10,-2,-10,-5,20], k = 2
//   Output: 23
//   Explanation: The subsequence is [10, -2, -5, 20].
// Constraints:
//   1 <= k <= nums.length <= 10^5
//   -10^4 <= nums[i] <= 10^4

func constrainedSubsetSum(nums []int, k int) int {
    // we can always include positive number, for negative number,
    // we need dp to check the max value we can get
    dp := make([]int, len(nums))   // dp[i] is the max sum for nums[i:] with nums[i] included
    dp[len(nums)-1] = nums[len(nums)-1]
	allmax := nums[len(nums)-1]
	l := list.New()
	l.PushBack(len(nums)-1)

	// first get max for dp[i+1], ..., dp[i+k]
	// then check if we need pop the queue (l.front-i == k)
	// then add i to queue back, before add, delete all elements that less than dp[i]
    for i:=len(nums)-2; i>=0; i--{
    	// max(dp[i+1], ..., dp[i+k]) = dp[l.front]
		// dp[i]=nums[i]+max(dp[i+1], dp[i+2], ..., dp[i+k], 0)
		dp[i] = nums[i] + max(dp[l.Front().Value.(int)], 0)
    	if l.Len()>0 && l.Front().Value.(int)-i == k {      // l.front-i == k
    		l.Remove(l.Front())
		}
		for l.Len()>0 && dp[l.Back().Value.(int)] < dp[i] {
			l.Remove(l.Back())
		}
		l.PushBack(i)
		if dp[i]>allmax {
			allmax = dp[i]
		}
	}
	return allmax
}

func max(a, b int) int {
	if a>b {
		return a
	}
	return b
}

func main() {
	fmt.Println(constrainedSubsetSum([]int{10,2,-10,5,20}, 2))
	fmt.Println(constrainedSubsetSum([]int{-1,-2,-3}, 1))
	fmt.Println(constrainedSubsetSum([]int{-3,-1,-2}, 1))
	fmt.Println(constrainedSubsetSum([]int{10,-2,-10,-5,20}, 1))
	fmt.Println(constrainedSubsetSum([]int{10,-2,-10,-5,20}, 2))
}

File: gotest/leet_rust/tests/leet1425_constrained_subsequence_sum.rs
// https://leetcode.com/problems/constrained-subsequence-sum/

// Given an integer array nums and an integer k, return the maximum sum of a non-empty subsequence
// of that array such that for every two consecutive integers in the subsequence, nums[i] and nums[j],
// where i < j, the condition j - i <= k is satisfied.
// A subsequence of an array is obtained by deleting some number of elements (can be zero) from the
// array, leaving the remaining elements in their original order.
// Example 1:
//   Input: nums = [10,2,-10,5,20], k = 2
//   Output: 37
//   Explanation: The subsequence is [10, 2, 5, 20].
// Example 2:
//   Input: nums = [-1,-2,-3], k = 1
//   Output: -1
//   Explanation: The subsequence must be non-empty, so we choose the largest number.
// Example 3:
//   Input: nums = [10,-2,-10,-5,20], k = 2
//   Output: 23
//   Explanation: The subsequence is [10, -2, -5, 20].
// Constraints:
//   1 <= k <= nums.length <= 10⁵
//   -10⁴ <= nums[i] <= 10⁴

mod _constrained_subsequence_sum {
    struct Solution{
        nums: Vec<i32>,
        k: i32,
        ans: i32,
    }

    // sliding window: let dp[i] be the maximum result you can get if the last element is nums[i],
    // then dp[i] = max(0, dp[i-k], ..., dp[i-1]) + nums[i]
    // use deque for sliding window minimal
    use std::collections::VecDeque;
    impl Solution {
        pub fn constrained_subset_sum(nums: Vec<i32>, k: i32) -> i32 {
            let mut ans = nums[0];
            let mut dp = vec![0; nums.len()];
            let k = k as usize;
            let mut deque = VecDeque::new();
            deque.push_back(0);
            dp[0] = nums[0];
            for i in 1..nums.len() {
                // get dp[i] = max(0, dp[i-k], ..., dp[i-1]) + nums[i]
                dp[i] = 0.max(dp[*deque.front().unwrap()]) + nums[i];
                ans = ans.max(dp[i]);
                // add dp[i] to queue
                if i-*deque.front().unwrap() >= k {
                    deque.pop_front();
                }
                while deque.len() > 0 && dp[*deque.back().unwrap()] <= dp[i] {
                    deque.pop_back();
                }
                deque.push_back(i);
            }
            return ans;
        }
    }

    #[test]
    fn test() {
        let testcases = vec![
            Solution {
                nums: vec![10,2,-10,5,20],
                k: 2,
                ans: 37,
            },
            Solution {
                nums: vec![-1,-2,-3],
                k: 1,
                ans: -1,
            },
            Solution {
                nums: vec![10,-2,-10,-5,20],
                k: 2,
                ans: 23,
            },
        ];
        for i in testcases {
            let ans = Solution::constrained_subset_sum(i.nums, i.k);
            println!("{}, {}", ans, i.ans);
        }
    } 
}